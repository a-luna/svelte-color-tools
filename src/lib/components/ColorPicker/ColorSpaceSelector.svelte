<script lang="ts">
	import Select from '$lib/components/Select/Select.svelte';
	import type { ColorSpace, SelectMenuOption } from '$lib/types';

	export let width = '100%';
	export let fontSize: string = '0.875rem';
	export let value: string = 'rgb';

	const options: SelectMenuOption[] = [
		{ label: 'RGB', value: 'rgb', optionNumber: 1, active: false },
		{ label: 'RGBA', value: 'rgba', optionNumber: 2, active: false },
		{ label: 'HSL', value: 'hsl', optionNumber: 3, active: false },
		{ label: 'HSLA', value: 'hsla', optionNumber: 4, active: false },
	];
	const menuId = 'select-color-space';
	const menuLabel = '';

	function handleColorSpaceChanged(colorSpace: ColorSpace) {
		value = colorSpace;
	}
</script>

<Select
	{menuLabel}
	{options}
	selectedValue={value}
	{menuId}
	{width}
	{fontSize}
	on:changed={(e) => handleColorSpaceChanged(e.detail)}
/>
