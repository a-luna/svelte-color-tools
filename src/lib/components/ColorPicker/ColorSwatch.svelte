<script lang="ts">
	import { hslToString } from '$lib/color';
	import ColorSwatches from '$lib/components/Icons/ColorSwatches.svelte';
	import { getColorPickerStore } from '$lib/context';
	import { createEventDispatcher } from 'svelte';

	export let pickerId: string;
	let state = getColorPickerStore(pickerId);
	const dispatch = createEventDispatcher();

	$: alphaBg = `background-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='8' height='8' viewBox='0 0 8 8'%3E%3Cg fill='%23000000' fill-opacity='0.4'%3E%3Cpath fill-rule='evenodd' d='M0 0h4v4H0V0zm4 4h4v4H4V4z'/%3E%3C/g%3E%3C/svg%3E"); background-size: 20px 20px; background-position: 0 0, 0 10px, 10px -10px, -10px 0px;`;
	$: swatchBgStyles = $state.alphaEnabled ? alphaBg : 'background-color: inherit';
	$: bgColor = `background-color: ${$state?.color?.hslaString};`;
	$: borderColor = { ...$state?.color?.hsl, l: $state?.color?.hsl.l - 20 };
	$: borderStyles = `border: 2px solid ${hslToString(borderColor)};`;
	$: pointerStyles = $state.editable ? '' : `pointer-events: none`;
	$: tooltip = $state.alphaEnabled
		? 'Color picker does not support RGBA/HSLA color space'
		: 'Click to open color picker';

	function showColorPicker() {
		if (!$state.alphaEnabled && $state.editable) {
			$state.labelState = 'pick';
			dispatch('showColorPicker');
		}
	}
</script>

<div
	class="swatch-wrapper flex-grow"
	class:cursor-pointer={!$state.alphaEnabled && $state?.editable}
	class:cursor-not-allowed={$state.alphaEnabled || !$state?.editable}
	style="width: 104px; height: 104px; {borderStyles} {pointerStyles}"
	title={tooltip}
	on:click={() => showColorPicker()}
>
	<div class="swatch" style={swatchBgStyles} />
	<div class="swatch-overlay" style={bgColor} />
	<div class="icon" title="Open X11 Color Palettes" on:click={() => dispatch('showX11Palettes')}>
		<ColorSwatches />
	</div>
</div>

<style lang="postcss">
	.swatch-wrapper {
		display: grid;
		grid-template-columns: 1fr;
		grid-template-rows: 1fr 25px;
		border-radius: 6px;
	}

	.swatch {
		z-index: 1;
		position: relative;
		border-radius: 4px;

		grid-column: 1 / span 1;
		grid-row: 1 / span 2;
	}

	.swatch-overlay {
		z-index: 2;
		border-radius: 4px;

		grid-column: 1 / span 1;
		grid-row: 1 / span 2;
	}

	.icon {
		z-index: 3;
		cursor: pointer;
		color: var(--black4);
		width: 25px;
		height: 25px;
		padding: 0 0 0.25rem 0.25rem;

		grid-column: 1 / span 1;
		grid-row: 2 / span 1;
	}
</style>
