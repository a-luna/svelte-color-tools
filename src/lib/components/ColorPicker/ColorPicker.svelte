<svelte:options accessors />

<script lang="ts">
	import { getX11ColorPalettes } from '$lib/color';
	import ColorChannels from '$lib/components/ColorPicker/ColorChannels/ColorChannels.svelte';
	import ColorLabel from '$lib/components/ColorPicker/ColorLabel/ColorLabel.svelte';
	import ColorSpaceSelector from '$lib/components/ColorPicker/ColorSpaceSelector.svelte';
	import ColorSwatch from '$lib/components/ColorPicker/ColorSwatch.svelte';
	import X11Palettes from '$lib/components/ThemeEditor/Palettes/X11Palettes/X11Palettes.svelte';
	import { getRandomHexString } from '$lib/helpers';
	import { ColorParser } from '$lib/parser';
	import type { ColorPalette, ColorPickerState, CssColor } from '$lib/types';
	import { onDestroy, onMount, setContext, tick } from 'svelte';
	import { writable } from 'svelte/store';

	export let pickerId: string = `color-picker-${getRandomHexString(4)}`;
	export let state = writable<ColorPickerState>({
		pickerId,
		color: ColorParser.parse('rgb(128 128 128)').value,
		colorSpace: 'rgb',
		labelState: 'prerender',
		editable: true,
	});

	export let editable = true;
	export let showX11Palettes = false;
	setContext($state.pickerId, { state });
	let timeout: NodeJS.Timeout;
	let colorPicker: HTMLInputElement;
	let x11ColorPalettes: ColorPalette[];

	$: alphaEnabled = $state.colorSpace === 'rgba' || $state.colorSpace === 'hsla';
	$: $state.editable = editable;
	$: if (typeof window !== 'undefined') x11ColorPalettes = getX11ColorPalettes();

	export function setColor(color: CssColor) {
		setCorrectColorSpace(color);
		$state.color = color;
		$state.labelState = 'success';
		timeout = setTimeout(() => {
			$state.labelState = 'inactive';
		}, 500);
	}

	function setCorrectColorSpace(color: CssColor) {
		if ($state.colorSpace === 'rgba' && !color.hasAlpha) {
			$state.colorSpace = 'rgb';
		} else if ($state.colorSpace === 'rgb' && color.hasAlpha) {
			$state.colorSpace = 'rgba';
		} else if ($state.colorSpace === 'hsla' && !color.hasAlpha) {
			$state.colorSpace = 'hsl';
		} else if ($state.colorSpace === 'hsl' && color.hasAlpha) {
			$state.colorSpace = 'hsla';
		}
	}

	const getHexOpaqueValue = (color: CssColor): string => color?.hex?.slice(0, 7) || '#000000';

	function handleStringValueChanged(color: string) {
		const result = ColorParser.parse(color);
		if (result.success) {
			$state.color = result.value;
			$state.labelState = 'success';
		} else {
			$state.labelState = 'error';
		}
		timeout = setTimeout(() => {
			$state.labelState = 'inactive';
		}, 500);
	}

	function handleColorPickerValueChanged() {
		$state.color = ColorParser.parse(colorPicker.value).value;
		$state.labelState = 'success';
		timeout = setTimeout(() => {
			$state.labelState = 'inactive';
		}, 500);
	}

	onMount(async () => {
		await tick();
		$state.labelState = 'inactive';
	});

	onDestroy(() => clearTimeout(timeout));
</script>

<input
	bind:this={colorPicker}
	type="color"
	style="display: none"
	value={getHexOpaqueValue($state?.color)}
	on:change={() => handleColorPickerValueChanged()}
/>
{#if !showX11Palettes}
	<div class="color-picker" data-testid={$state.pickerId}>
		<div class="picker-left-col">
			<ColorSpaceSelector bind:value={$state.colorSpace} disabled={!$state.editable} />
			<ColorSwatch
				pickerId={$state.pickerId}
				{alphaEnabled}
				on:showX11Palettes={() => (showX11Palettes = true)}
				on:showColorPicker={() => colorPicker.click()}
			/>
		</div>
		<div class="picker-right-col">
			<ColorLabel
				pickerId={$state.pickerId}
				{alphaEnabled}
				on:updateColor={(e) => handleStringValueChanged(e.detail)}
			/>
			<ColorChannels pickerId={$state.pickerId} {alphaEnabled} editable={$state.editable} />
		</div>
	</div>
{:else if x11ColorPalettes !== undefined}
	<X11Palettes
		{x11ColorPalettes}
		on:colorSelected={() => (showX11Palettes = false)}
		on:colorSelected={(e) => setColor(e.detail)}
	/>
{/if}

<style lang="postcss">
	.color-picker {
		display: flex;
		flex-flow: row nowrap;
		align-items: flex-start;
		gap: 0.5rem;
		width: min-content;
		background-color: var(--white1);
		border-radius: 4px;
		padding: 0.5rem;
	}

	.picker-left-col,
	.picker-right-col {
		display: flex;
		flex-flow: column nowrap;
		justify-content: flex-start;
		align-items: stretch;
		gap: 0.5rem;
	}
</style>
