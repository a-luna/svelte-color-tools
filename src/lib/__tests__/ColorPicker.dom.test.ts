import ColorPicker from '$lib/components/ColorPicker/ColorPicker.svelte';
import { fireEvent, render, screen } from '@testing-library/svelte';

describe('ColorPicker component can parse strings containing valid color keywords (CSS Colors Level 3)', () => {
	it('can parse color name, regardless of letter-case: fuchsia/Fuchsia/FUCHSIA', async () => {
		const pickerId = 'test-picker';
		const { container, getByTestId } = render(ColorPicker, {
			pickerId,
		});

		expect(container).toMatchSnapshot('initial-state');
		const picker = await screen.findByTestId(pickerId);
		expect(picker).toHaveClass('color-picker');

		let colorString = await screen.findByTestId('color-string');
		expect(colorString).toHaveTextContent('#808080');
		await fireEvent.click(colorString);
		expect(colorString).toHaveTextContent('rgb(128 128 128)');
		await fireEvent.click(colorString);
		expect(colorString).toHaveTextContent('hsl(0 0% 50%)');

		const redSlider = await screen.findByTestId('R-slider');
		expect(redSlider).toHaveValue('128');
		fireEvent.change(redSlider, { target: { value: '0' } });
		const greenSlider = await screen.findByTestId('G-slider');
		expect(greenSlider).toHaveValue('128');
		fireEvent.change(greenSlider, { target: { value: '0' } });
		const blueSlider = await screen.findByTestId('B-slider');
		expect(blueSlider).toHaveValue('128');
		fireEvent.change(blueSlider, { target: { value: '255' } });
		expect(redSlider).toHaveValue('0');
		expect(greenSlider).toHaveValue('0');
		expect(blueSlider).toHaveValue('255');

		const colorSpaceSelectorButton = await screen.findByTestId('select-color-space-open-list-button');
		await fireEvent.click(colorSpaceSelectorButton);
		const hslaMenuOption = await screen.findByTestId('select-color-space-option-4');
		expect(hslaMenuOption).toBeVisible();
		await fireEvent.click(hslaMenuOption);

		const hueSlider = await screen.findByTestId('H-slider');
		expect(hueSlider).toHaveValue('240');
		fireEvent.change(hueSlider, { target: { value: '290' } });
		const satSlider = await screen.findByTestId('S-slider');
		expect(satSlider).toHaveValue('100');
		const lightSlider = await screen.findByTestId('L-slider');
		expect(lightSlider).toHaveValue('50');
		fireEvent.change(lightSlider, { target: { value: '75' } });
		const alphaSlider = await screen.findByTestId('A-slider');
		expect(alphaSlider).toHaveValue('1');
		fireEvent.change(alphaSlider, { target: { value: '0.44' } });

		expect(hueSlider).toHaveValue('290');
		expect(satSlider).toHaveValue('100');
		expect(lightSlider).toHaveValue('75');
		expect(alphaSlider).toHaveValue('0.44');

		let editButton = await screen.findByTestId('edit-color-string-button');
		await fireEvent.click(editButton);
		let colorNameInput = await screen.findByTestId('color-name-input');
		await fireEvent.change(colorNameInput, { target: { value: 'fuchsia' } });
		expect(colorNameInput).toHaveValue('fuchsia');
		let okButton = await screen.findByTestId('change-color-button');
		await fireEvent.click(okButton);

		colorString = await screen.findByTestId('color-string');
		expect(colorString).toHaveTextContent('#ff00ff70');
		const copyButton = await screen.findByTestId('copy-color-string-button');
		await fireEvent.click(copyButton);

		editButton = await screen.findByTestId('edit-color-string-button');
		await fireEvent.click(editButton);
		colorNameInput = await screen.findByTestId('color-name-input');
		await fireEvent.change(colorNameInput, { target: { value: 'Fuchsia' } });
		expect(colorNameInput).toHaveValue('Fuchsia');
		okButton = await screen.findByTestId('change-color-button');
		await fireEvent.click(okButton);

		colorString = await screen.findByTestId('color-string');
		expect(colorString).toHaveTextContent('#ff00ff70');

		editButton = await screen.findByTestId('edit-color-string-button');
		await fireEvent.click(editButton);
		colorNameInput = await screen.findByTestId('color-name-input');
		await fireEvent.change(colorNameInput, { target: { value: 'FUCHSIA' } });
		expect(colorNameInput).toHaveValue('FUCHSIA');
		okButton = await screen.findByTestId('change-color-button');
		await fireEvent.click(okButton);

		colorString = await screen.findByTestId('color-string');
		expect(colorString).toHaveTextContent('#ff00ff70');
	});

	it('can parse color name, regarless of spaces, dashes or underscores: lawngreen/LawnGreen/lawn-green/lawn green/Lawn_Green', async () => {
		const pickerId = 'test-picker';
		const { container, getByTestId } = render(ColorPicker, {
			pickerId,
		});

		let editButton = await screen.findByTestId('edit-color-string-button');
		await fireEvent.click(editButton);
		let colorNameInput = await screen.findByTestId('color-name-input');
		await fireEvent.change(colorNameInput, { target: { value: 'lawngreen' } });
		expect(colorNameInput).toHaveValue('lawngreen');
		let okButton = await screen.findByTestId('change-color-button');
		await fireEvent.click(okButton);

		let colorString = await screen.findByTestId('color-string');
		expect(colorString).toHaveTextContent('#7cfc00');

		editButton = await screen.findByTestId('edit-color-string-button');
		await fireEvent.click(editButton);
		colorNameInput = await screen.findByTestId('color-name-input');
		await fireEvent.change(colorNameInput, { target: { value: 'LawnGreen' } });
		expect(colorNameInput).toHaveValue('LawnGreen');
		okButton = await screen.findByTestId('change-color-button');
		await fireEvent.click(okButton);

		colorString = await screen.findByTestId('color-string');
		expect(colorString).toHaveTextContent('#7cfc00');

		editButton = await screen.findByTestId('edit-color-string-button');
		await fireEvent.click(editButton);
		colorNameInput = await screen.findByTestId('color-name-input');
		await fireEvent.change(colorNameInput, { target: { value: 'lawn-green' } });
		expect(colorNameInput).toHaveValue('lawn-green');
		okButton = await screen.findByTestId('change-color-button');
		await fireEvent.click(okButton);

		colorString = await screen.findByTestId('color-string');
		expect(colorString).toHaveTextContent('#7cfc00');

		editButton = await screen.findByTestId('edit-color-string-button');
		await fireEvent.click(editButton);
		colorNameInput = await screen.findByTestId('color-name-input');
		await fireEvent.change(colorNameInput, { target: { value: 'lawn green' } });
		expect(colorNameInput).toHaveValue('lawn green');
		okButton = await screen.findByTestId('change-color-button');
		await fireEvent.click(okButton);

		colorString = await screen.findByTestId('color-string');
		expect(colorString).toHaveTextContent('#7cfc00');

		editButton = await screen.findByTestId('edit-color-string-button');
		await fireEvent.click(editButton);
		colorNameInput = await screen.findByTestId('color-name-input');
		await fireEvent.change(colorNameInput, { target: { value: 'Lawn_Green' } });
		expect(colorNameInput).toHaveValue('Lawn_Green');
		okButton = await screen.findByTestId('change-color-button');
		await fireEvent.click(okButton);

		colorString = await screen.findByTestId('color-string');
		expect(colorString).toHaveTextContent('#7cfc00');
	});
});
