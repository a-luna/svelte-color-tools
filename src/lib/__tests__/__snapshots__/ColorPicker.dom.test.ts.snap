// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorPicker component can parse strings containing valid color keywords (CSS Colors Level 3) can parse color name, regardless of letter-case: fuchsia/Fuchsia/FUCHSIA: initial-state 1`] = `
<body>
  <div>
    <input
      style="display: none;"
      type="color"
    />
     
    <div
      class="color-picker svelte-1isu17t"
      data-testid="test-picker"
    >
      <div
        class="picker-left-col svelte-1isu17t"
      >
        <div
          class="relative inline-block text-left flex-initial svelte-1sobnzl"
          data-testid="select-color-space"
          style="width: 100%; margin: 0px;"
        >
          <div>
            <button
              aria-expanded="false"
              aria-haspopup="true"
              class="font-medium inline-flex items-center justify-between gap-2.5 w-full border-0 rounded-md shadow-sm focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-offset-gray-100 focus:ring-gray-800 open-list-button svelte-1sobnzl"
              data-testid="select-color-space-open-list-button"
              id="select-color-space-open-list-button"
              style="font-size: 0.875rem; height: 36px; padding: 10px 8px;"
              type="button"
            >
              <span
                class="leading-none whitespace-nowrap mx-auto"
              >
                RGB
              </span>
               
              <div
                style="width: 0.875rem; height: 0.875rem;"
              >
                <svg
                  class="svelte-q1wtpt"
                  viewBox="0 0 320 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  
                  <path
                    d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                  />
                </svg>
              </div>
            </button>
          </div>
           
        </div>
         
        <div
          class="swatch-wrapper flex-grow svelte-1cmbpwl cursor-pointer"
          style="background-color: hsl(0 0% 50%); width: 104px; height: 104px; border: 2px solid hsla(0 0% 30% / 1)"
          title="Click to open color picker"
        >
          <div
            class="swatch svelte-1cmbpwl"
            style="background-color: hsl(0 0% 50%)"
          />
           
          <div
            class="icon svelte-1cmbpwl"
            title="Click to view X11 color palettes"
          >
            <svg
              aria-hidden="true"
              class="svelte-1ebfv77"
              focusable="false"
              style="width: 0px; height: 0px; position: absolute;"
            >
              <lineargradient
                id="gradient-horizontal"
                x2="1"
                y2="1"
              >
                <stop
                  offset="0%"
                  stop-color="#0044FF"
                />
                <stop
                  offset="50%"
                  stop-color="#76E0C4"
                />
                <stop
                  offset="100%"
                  stop-color="#8CFF1A"
                />
              </lineargradient>
              <lineargradient
                id="gradient-hover"
                x2="1"
                y2="1"
              >
                <stop
                  offset="0%"
                  stop-color="#0036CC"
                />
                <stop
                  offset="50%"
                  stop-color="#4CD6B2"
                />
                <stop
                  offset="100%"
                  stop-color="#73E600"
                />
              </lineargradient>
            </svg>
             
            <svg
              class="svelte-1ebfv77"
              viewBox="0 0 511 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M479.06 320H372.29L186.15 506.51c-2.06 2.07-4.49 3.58-6.67 5.49h299.58c17.64 0 31.94-14.33 31.94-32V352c0-17.67-14.3-32-31.94-32zm-44.5-152.9l-90.33-90.51c-12.47-12.5-32.69-12.5-45.17 0l-75.5 75.65V416c0 2.96-.67 5.73-.87 8.64l211.87-212.28c12.47-12.5 12.47-32.77 0-45.26zM191.62 32c0-17.67-14.3-32-31.94-32H31.94C14.3 0 0 14.33 0 32v384c0 53.02 42.9 96 95.81 96s95.81-42.98 95.81-96V32zM95.81 440c-13.23 0-23.95-10.75-23.95-24 0-13.26 10.73-24 23.95-24s23.95 10.74 23.95 24c.01 13.25-10.72 24-23.95 24zm31.94-184H63.88v-64h63.88v64zm0-128H63.88V64h63.88v64z"
              />
            </svg>
          </div>
        </div>
      </div>
       
      <div
        class="picker-right-col svelte-1isu17t"
      >
        <div
          class="color-label p-2.5 svelte-vvf2yt"
          style="background-color: var(--white4)"
        >
          <button
            class="flex-initial h-4 w-4 my-auto cursor-pointer svelte-15rom32"
            data-testid="copy-color-string-button"
            id="copy-color-string-button"
            title="Copy Hex value"
            type="button"
          >
            <svg
              class="svelte-q1wtpt"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              
              <path
                d="M433.941 65.941l-51.882-51.882A48 48 0 0 0 348.118 0H176c-26.51 0-48 21.49-48 48v48H48c-26.51 0-48 21.49-48 48v320c0 26.51 21.49 48 48 48h224c26.51 0 48-21.49 48-48v-48h80c26.51 0 48-21.49 48-48V99.882a48 48 0 0 0-14.059-33.941zM266 464H54a6 6 0 0 1-6-6V150a6 6 0 0 1 6-6h74v224c0 26.51 21.49 48 48 48h96v42a6 6 0 0 1-6 6zm128-96H182a6 6 0 0 1-6-6V54a6 6 0 0 1 6-6h106v88c0 13.255 10.745 24 24 24h88v202a6 6 0 0 1-6 6zm6-256h-64V48h9.632c1.591 0 3.117.632 4.243 1.757l48.368 48.368a6 6 0 0 1 1.757 4.243V112z"
              />
            </svg>
          </button>
           
          <span
            class="cursor-pointer flex-grow text-center self-center font-medium leading-none whitespace-nowrap"
            data-testid="color-string"
            style="font-size: 0.85rem;"
            title="Click to toggle Hex, RGB, and HSL string values"
          >
            #808080
          </span>
           
          <button
            class="flex-initial h-4 w-4 my-auto svelte-15rom32 cursor-pointer"
            data-testid="edit-color-string-button"
            id="edit-color-string-button"
            title="Edit string value"
            type="button"
          >
            <svg
              class="svelte-q1wtpt"
              viewBox="0 0 576 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              
              <path
                d="M402.3 344.9l32-32c5-5 13.7-1.5 13.7 5.7V464c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V112c0-26.5 21.5-48 48-48h273.5c7.1 0 10.7 8.6 5.7 13.7l-32 32c-1.5 1.5-3.5 2.3-5.7 2.3H48v352h352V350.5c0-2.1.8-4.1 2.3-5.6zm156.6-201.8L296.3 405.7l-90.4 10c-26.2 2.9-48.5-19.2-45.6-45.6l10-90.4L432.9 17.1c22.9-22.9 59.9-22.9 82.7 0l43.2 43.2c22.9 22.9 22.9 60 .1 82.8zM460.1 174L402 115.9 216.2 301.8l-7.3 65.3 65.3-7.3L460.1 174zm64.8-79.7l-43.2-43.2c-4.1-4.1-10.8-4.1-14.8 0L436 82l58.1 58.1 30.9-30.9c4-4.2 4-10.8-.1-14.9z"
              />
            </svg>
          </button>
        </div>
         
        <div
          class="text-sm flex flex-col flex-nowrap justify-center items-stretch gap-2"
        >
          <div
            class="flex flex-row flex-nowrap items-center gap-2 svelte-1tdtogt"
          >
            <label
              class="text-right flex-initial font-medium svelte-1tdtogt"
              for="slider"
            >
              R
            </label>
             
            <input
              class="m-0 flex-grow svelte-1tdtogt"
              data-testid="R-slider"
              id="R-slider"
              max="255"
              min="0"
              step="1"
              type="range"
            />
             
            <span
              class="value flex-initial font-medium svelte-1tdtogt"
            >
              128
            </span>
          </div>
           
          <div
            class="flex flex-row flex-nowrap items-center gap-2 svelte-1tdtogt"
          >
            <label
              class="text-right flex-initial font-medium svelte-1tdtogt"
              for="slider"
            >
              G
            </label>
             
            <input
              class="m-0 flex-grow svelte-1tdtogt"
              data-testid="G-slider"
              id="G-slider"
              max="255"
              min="0"
              step="1"
              type="range"
            />
             
            <span
              class="value flex-initial font-medium svelte-1tdtogt"
            >
              128
            </span>
          </div>
           
          <div
            class="flex flex-row flex-nowrap items-center gap-2 svelte-1tdtogt"
          >
            <label
              class="text-right flex-initial font-medium svelte-1tdtogt"
              for="slider"
            >
              B
            </label>
             
            <input
              class="m-0 flex-grow svelte-1tdtogt"
              data-testid="B-slider"
              id="B-slider"
              max="255"
              min="0"
              step="1"
              type="range"
            />
             
            <span
              class="value flex-initial font-medium svelte-1tdtogt"
            >
              128
            </span>
          </div>
           
          <div
            style="width: 100%; height: 20px;"
          />
          
        </div>
      </div>
    </div>
    
  </div>
</body>
`;
